rubocop-klaxit/spec/cops/no_undefined_variable_usage_spec.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::NoUndefinedVariableUsage, :config do
+  it "registers an offense for simple self assignments" do
+    expect_offense(<<~RUBY)
+      foo = foo
+      ^^^^^^^^^ Variable used before being defined
+    RUBY
+  end
+  it "registers an offense for simple assignments with multiple nodes" do
+    expect_offense(<<~RUBY)
+      foo = 1 + foo - 3
+      ^^^^^^^^^^^^^^^^^ Variable used before being defined
+    RUBY
+  end
+  it "registers an offense for assignments using variable methods" do
+    expect_offense(<<~RUBY)
+      foo = foo.increment
+      ^^^^^^^^^^^^^^^^^^^ Variable used before being defined
+    RUBY
+  end
+  it "registers an offense for assignments which override method name" do
+    expect_offense(<<~RUBY)
+      def foo
+        1
+      end
+
+      foo = foo
+      ^^^^^^^^^ Variable used before being defined
+    RUBY
+  end
+  it "registers an offense for assignments using variable in a block" do
+    expect_offense(<<~RUBY)
+      foo = begin
+      ^^^^^^^^^^^ Variable used before being defined
+        foo
+      end
+    RUBY
+  end
+  it "registers an offense for multiple self assignments" do
+    expect_offense(<<~RUBY)
+      foo, bar = foo, bar
+      ^^^ Variable used before being defined
+           ^^^ Variable used before being defined
+    RUBY
+  end
+  it "registers an offense for assignments using definitions outside scope" do
+    expect_offense(<<~RUBY)
+      foo = 1
+      def bar
+        foo = foo
+        ^^^^^^^^^ Variable used before being defined
+      end
+    RUBY
+  end
+  it "registers an offense for problematic assignments in parent contexts" do
+    expect_offense(<<~RUBY)
+      foo = foo
+      ^^^^^^^^^ Variable used before being defined
+      begin
+        foo = foo
+        ^^^^^^^^^ Variable used before being defined
+        begin
+          foo = foo
+          ^^^^^^^^^ Variable used before being defined
```

Comment:
Isn't it weird that the offense is registered n times?

Reply:
It is, but since we are using `on_lvasgn`, we'd have to have some sort of mechanism which remembers if it has already been registered, or only signal the outermost context each time (this would require changing the logic a fair bit). I think it's fine leaving it as it is for now, and if this cop is used extensively as cases like this are supposed to be quite rare.. (hopefully)

Reply:
@teckwan just thinking a bit about it: we could however remove that particular test, since having a regression here would be a benefit!

Reply:
signaling only outermost context was easier than expected, and I fixed an error. @BuonOmo

Reply:
Well done :clap:

I'll let @Quiwin be the second reviewer